ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/stats/HiveRelMdDistinctRowCount.java

```diff
@@ -93,6 +89,16 @@ public Double getDistinctRowCount(HiveJoin rel, RelMetadataQuery mq, ImmutableBi
            groupKey, predicate, true);
   }
 
+  public Double getDistinctRowCount(RelNode r, RelMetadataQuery mq, ImmutableBitSet groupKey,
+                                    RexNode predicate) {
+    if (r instanceof SingleRel) {
+      return mq.getDistinctRowCount(r.getInput(0), groupKey, predicate);
+    } else if (r instanceof JdbcHiveTableScan) {
+      return getDistinctRowCount(((JdbcHiveTableScan) r).getHiveTableScan(), mq, groupKey, predicate);
+    }
+
+    return super.getDistinctRowCount(r, mq, groupKey, predicate);
+  }
```

Comment:
Can you explain a bit what's the general idea behind this change cause the logic looks a bit brittle. 

`SingleRel` has many sub-classes and some of them can greatly affect the row count so saying that you should get the row count by considering only its input doesn't look right.

Regarding the `r istanceof JdbcHiveTableScan` check I think it would be better to add a separate method `getDistictRowCount(JdbcHiveTableScan...)` and do exactly what you did here.

Reply:
Added a comment in the method with a brief explanation. Changed `SingleRel` to `Converter`, and added a method for `getDistictRowCount(JdbcHiveTableScan...)`.

Reply:
Is it possible to remove this method entirely and add the following:
```
public Double getDistinctRowCount(Converter rel, RelMetadataQuery mq, ImmutableBitSet groupKey,
                                    RexNode predicate) {
}
```

Reply:
A little late, but this is done now!